In [1]:
import pandas as pd


In [ ]:
df = pd.read_csv("Titanic-Dataset.csv")  # or the exact filename you downloaded

df.head()
df.info()
df.describe()
df.isnull().sum()


In [5]:
df.shape


(891, 12)

In [6]:
df.info()
df.isnull().sum()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

We have 2 float columns and 5 int columns and 5 object columns

In [9]:
df["Survived"].value_counts()

df["Survived"].value_counts(normalize=True) * 100



Survived
0    61.616162
1    38.383838
Name: proportion, dtype: float64

“X passengers survived (38%), Z passengers did not (62%).”

In [10]:
df["Sex"].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [11]:
df.groupby("Sex")["Survived"].mean() * 100

Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64

In [12]:
df["Pclass"].value_counts()


Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [13]:
df.groupby("Pclass")["Survived"].mean() * 100

Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64

“Survival rates by class: 1st: 67%, 2nd: 48%, 3rd: 25%.”



In [14]:
df["Age"].mean()


np.float64(29.69911764705882)

In [15]:
young = df[df["Age"] < 18]
len(young)


113

In [17]:
young["Survived"].mean() * 100


np.float64(53.98230088495575)

“There were N passengers under 18, with a survival rate of 54%.”



SECTION 2

In [18]:
df["FamilySize"] = df["SibSp"] + df["Parch"] + 1
df[["SibSp", "Parch", "FamilySize"]].head()

SibSp  Parch  FamilySize
0      1      0           2
1      1      0           2
2      0      0           1
3      1      0           2
4      0      0           1

SibSp = siblings/spouses aboard.

Parch = parents/children aboard.

FamilySize = total people in family including the passenger

In [19]:
df["IsAlone"] = 0
df.loc[df["FamilySize"] == 1, "IsAlone"] = 1
df[["FamilySize", "IsAlone"]].head()

FamilySize  IsAlone
0           2        0
1           2        0
2           1        1
3           2        0
4           1        1

df.loc[df["FamilySize"] == 1, "IsAlone"] = 1
df["FamilySize"] == 1 → a True/False mask for each row.

df.loc[mask, "IsAlone"] → “go to rows where the mask is True, and the column IsAlone”.

= 1 → set IsAlone to 1 there.

So .loc here means: “on these rows (FamilySize == 1), in this column (IsAlone), assign 1.”

In [20]:
df["FamilySize"].value_counts().sort_index()

FamilySize
1     537
2     161
3     102
4      29
5      15
6      22
7      12
8       6
11      7
Name: count, dtype: int64

In [21]:
df.groupby("FamilySize")["Survived"].mean() * 100

FamilySize
1     30.353818
2     55.279503
3     57.843137
4     72.413793
5     20.000000
6     13.636364
7     33.333333
8      0.000000
11     0.000000
Name: Survived, dtype: float64

In [22]:
df.groupby("IsAlone")["Survived"].mean() * 100

IsAlone
0    50.564972
1    30.353818
Name: Survived, dtype: float64

People with family size of 1 had 33% survival rate, while people with family size of 2 had 55% survival rate.  

People who were alone had a survival rate of 30% and those who were not alone had a survival rate of 50%.

Here’s a tight revision of the new things you learned in this step.

FamilySize and IsAlone
FamilySize:

python
df["FamilySize"] = df["SibSp"] + df["Parch"] + 1
Adds a new column counting how many family members are on board with the passenger (siblings/spouse + parents/children + themselves).
​

IsAlone:

python
df["IsAlone"] = 0
df.loc[df["FamilySize"] == 1, "IsAlone"] = 1
Start with IsAlone = 0 for everyone.

For rows where FamilySize == 1, set IsAlone = 1 (travelling alone).
​

.loc (label/condition selector)
General form:

python
df.loc[row_selector, column_selector]
row_selector: which rows (labels or a boolean condition).

column_selector: which columns (name or list of names).
​

Example you used:

python
df.loc[df["FamilySize"] == 1, "IsAlone"] = 1
df["FamilySize"] == 1 → True/False mask.

.loc[mask, "IsAlone"] → select “rows where mask is True, column IsAlone”.

= 1 → assign 1 there.

Meaning: “For passengers whose FamilySize is 1, mark IsAlone as 1.”

groupby with new features
You also used groupby with these new columns:

python
df.groupby("FamilySize")["Survived"].mean() * 100
df.groupby("IsAlone")["Survived"].mean() * 100
Same idea as before:

Group the data by FamilySize or IsAlone.

Take the Survived column in each group.

.mean() * 100 → survival rate (%) inside that group, not percentages that must sum to 100 across groups.
​